File: plugin/com.hello.uxml.tools.ui/src/com/hello/uxml/tools/framework/DropDownButton.java
package com.hello.uxml.tools.framework;

import com.hello.uxml.tools.framework.events.MouseEventArgs;

import java.util.EnumSet;

/**
 * Implements button that shows popup as drop down on mouse down.
 *
 * @author ferhat@
 */
public class DropDownButton extends Button {

  /** IsOpen property definition */
  public static PropertyDefinition isOpenPropDef = PropertySystem.register("IsOpen",
      Boolean.class, DropDownButton.class,
      new PropertyData(false, EnumSet.of(PropertyFlags.None)));

  /** Popup property definition */
  public static PropertyDefinition popupPropDef = PropertySystem.register("Popup",
      UIElement.class, DropDownButton.class,
      new PropertyData(null, EnumSet.of(PropertyFlags.None)));

  /**
   * Returns whether dropdown is open.
   */
  public boolean getIsOpen() {
    return (Boolean) getProperty(isOpenPropDef);
  }

  /**
   * Sets drop down open state.
   */
  public void setIsOpen(boolean value) {
    setProperty(isOpenPropDef, value);
  }

  /**
   * Sets or returns popup element.
   */
  public UIElement getPopup() {
    return (UIElement) getProperty(popupPropDef);
  }

  public void setPopup(UIElement element) {
    setProperty(popupPropDef, element);
  }

  @Override
  protected void onMouseDown(MouseEventArgs e) {
    captureMouse();
    e.setHandled(true);
    setProperty(isPressedPropDef, true);
    setProperty(isOpenPropDef, !getIsOpen());
  }

  @Override
  protected void onMouseUp(MouseEventArgs e) {
    e.setHandled(true);
    setProperty(isPressedPropDef, false);
    releaseMouse();
    if (getIsMouseOver()) {
      onMouseClick(e);
    }
  }
}

File: framework/lib/drop_down_button.dart
part of uxml;

/**
 * Implements button that shows popup as drop down on mouse down.
 *
 * @author:ferhat@ (Ferhat Buyukkokten)
 */
class DropDownButton extends Button {
  /** IsOpen property definition */
  static PropertyDefinition isOpenProperty;
  /** Popup property definition */
  static PropertyDefinition popupProperty;
  static ElementDef dropdownbuttonElementDef;

  DropDownButton() : super() {
  }

  /**
   * Sets/returns open state.
   */
  bool get isOpen => getProperty(isOpenProperty);

  set isOpen(bool value) {
    return setProperty(isOpenProperty, value);
  }

  /**
   * Sets/returns popup element.
   */
  UIElement get popup => getProperty(popupProperty);

  set popup(UIElement value) {
    return setProperty(popupProperty, value);
  }

  /** Overrides Button.onMouseDown to toggle dropdown. */
  void onMouseDown(MouseEventArgs mouseArgs) {
    captureMouse();
    setProperty(Button.isPressedProperty, true);
    isOpen = !isOpen;
    mouseArgs.handled = true;
  }

  /** Overrides Button.onMouseUp. */
  void onMouseUp(MouseEventArgs mouseArgs) {
    releaseMouse();
    setProperty(Button.isPressedProperty, false);
    if (isMouseOver) {
      onClick(mouseArgs);
    }
  }

  static void _popupContentChanged(Object target, Object propDef,
      Object oldValue, Object newValue) {
    // TODO(ferhat): implement Popup hosting and wiring up isOpen properties.
  }

  /** @see UxmlElement.getDefinition. */
  ElementDef getDefinition() => dropdownbuttonElementDef;

  /** Registers component. */
  static void registerDropDownButton() {
    isOpenProperty = ElementRegistry.registerProperty("isOpen",
        PropertyType.BOOL, PropertyFlags.NONE, null, false);
    popupProperty = ElementRegistry.registerProperty("popup",
        PropertyType.UIELEMENT, PropertyFlags.NONE, _popupContentChanged, null);
    dropdownbuttonElementDef = ElementRegistry.register("DropDownButton",
        Button.buttonElementDef, [isOpenProperty, popupProperty], null);
  }
}
